File: src/mvmctl/cli/network.py
"""Network management commands."""

from __future__ import annotations

import json
from typing import TYPE_CHECKING

import typer
from rich.prompt import Prompt

from mvmctl.api import NetworkCreateInput as _NetworkCreateInput
from mvmctl.api import NetworkInput as _NetworkInput
from mvmctl.api import NetworkOperation as _NetworkOperation

if TYPE_CHECKING:
    from mvmctl.api.inputs._network_create_input import NetworkCreateInput
    from mvmctl.api.inputs._network_input import NetworkInput
    from mvmctl.api.network_operations import NetworkOperation
else:
    NetworkOperation = _NetworkOperation
    NetworkInput = _NetworkInput
    NetworkCreateInput = _NetworkCreateInput
from mvmctl.cli._completion import _complete_network_names
from mvmctl.models.result import OperationResult
from mvmctl.utils.cli import handle_errors, mvm_cli
from mvmctl.utils.network import NetworkUtils

if TYPE_CHECKING:
    from mvmctl.models import NetworkItem

network_app = typer.Typer(
    help="Network management",
    no_args_is_help=True,
    add_completion=False,
)


@network_app.callback()
def network_callback(ctx: typer.Context) -> None:
    pass


@network_app.command(name="help", hidden=True)
def help_cmd(ctx: typer.Context) -> None:
    """Show help for the network command group."""
    typer.echo(ctx.parent.get_help() if ctx.parent else "")
    raise typer.Exit()


@network_app.command(name="ls")
@handle_errors
def network_ls(
    json_output: bool = typer.Option(False, "--json", help="Output as JSON"),
) -> None:
    """List all networks."""
    networks: list[NetworkItem] = NetworkOperation.list_all()

    if json_output:
        data = NetworkOperation.to_json(networks)
        typer.echo(json.dumps(data, indent=2, default=str))
        return

    rows = []
    for n in networks:
        vm_count = len(n.leases) if n.leases else 0
        rows.append(
            [
                mvm_cli.format_marker(n.is_default),
                mvm_cli.format_id(n.id),
                mvm_cli.format_name(n.name, not n.is_present),
                n.subnet,
                n.bridge,
                "True" if n.nat_enabled else "False",
                str(vm_count),
                mvm_cli.format_timestamp(n.created_at),
            ]
        )
    mvm_cli.table(
        columns=[
            "",
            "ID",
            "Name",
            "Network",
            "Bridge",
            "NAT",
            "VMs",
            "Created",
        ],
        rows=rows,
    )


@network_app.command(
    name="default",
    context_settings={"allow_extra_args": True, "ignore_unknown_options": True},
)
@handle_errors
def network_set_default(
    ctx: typer.Context,
    name: str | None = typer.Argument(
        None,
        help="Network name to set as default",
        autocompletion=_complete_network_names,
    ),
) -> None:
    """Set a network as the default for VM creation."""
    name = mvm_cli.check_name_arg(ctx, name)
    result = NetworkOperation.set_default(NetworkInput(name=[name]))
    if result.status in ("error", "failure"):
        mvm_cli.error(result.message)
        raise typer.Exit(code=1)
    mvm_cli.success(f"Default network set to: {name}")


def _resolve_user_nat_gateways() -> str:
    """Interactive prompt for NAT gateway selection — CLI-only, no business logic."""
    interfaces = NetworkUtils.get_physical_interfaces()
    if not interfaces:
        mvm_cli.error("No network interfaces found")
        raise typer.Exit(code=1)
    if len(interfaces) == 1:
        return interfaces[0]

    mvm_cli.info("Select interface(s) for NAT (internet access):")
    for i, iface in enumerate(interfaces, 1):
        mvm_cli.info(f"  [{i}] {iface}")
    selected = Prompt.ask(
        "Select interface number(s) [comma-separated]", default="1"
    )
    try:
        indices = [int(x.strip()) for x in selected.split(",") if x.strip()]
        selected_interfaces = [
            interfaces[idx - 1]
            for idx in indices
            if 1 <= idx <= len(interfaces)
        ]
    except ValueError:
        mvm_cli.error(f"Invalid interface selection: {selected}")
        raise typer.Exit(code=1)
    if not selected_interfaces:
        mvm_cli.error("No valid interface indices selected")
        raise typer.Exit(code=1)
    return ",".join(selected_interfaces)


@network_app.command(
    name="create",
    context_settings={"allow_extra_args": True, "ignore_unknown_options": True},
)
@handle_errors
def network_create(
    ctx: typer.Context,
    name: str | None = typer.Argument(None, help="Network name"),
    subnet: str | None = typer.Option(
        None,
        "--subnet",
        help="IP subnet in SUBNET notation (e.g. 192.168.100.0/24)",
    ),
    ipv4_gateway: str | None = typer.Option(
        None, "--ipv4-gateway", help="Gateway IPv4 for the bridge"
    ),
    no_nat: bool = typer.Option(
        False, "--no-nat", help="Disable NAT/masquerade"
    ),
    nat_gateways: str | None = typer.Option(
        None,
        "--nat-gateways",
        help="Physical interfaces for NAT (comma-separated, auto-detected if not provided)",
    ),
    non_interactive: bool = typer.Option(
        False,
        "--non-interactive",
        help="Skip interactive prompts (auto-detect NAT interfaces)",
    ),
) -> None:
    """Create a named network."""
    name = mvm_cli.check_name_arg(ctx, name)

    if subnet is None:
        mvm_cli.error("Missing required option '--subnet'")
        raise typer.Exit(code=1)

    if nat_gateways is None and not no_nat and not non_interactive:
        nat_gateways = _resolve_user_nat_gateways()

    nat_gateways_list = (
        [g.strip() for g in nat_gateways.split(",") if g.strip()]
        if nat_gateways
        else []
    )

    create_input = NetworkCreateInput(
        name=name,
        subnet=subnet,
        ipv4_gateway=ipv4_gateway,
        nat_enabled=not no_nat,
        nat_gateways=nat_gateways_list,
    )
    result = NetworkOperation.create(create_input)
    if isinstance(result, OperationResult):
        if result.status in ("error", "failure"):
            mvm_cli.error(result.message)
            raise typer.Exit(code=1)
        if result.status == "skipped":
            mvm_cli.info(result.message)
            raise typer.Exit(code=0)
        config = result.item
        if config is None:
            mvm_cli.error("Network created but no item returned")
            raise typer.Exit(code=1)
    else:
        # NeedsInteraction — not expected for network creation
        mvm_cli.error(result.message)
        raise typer.Exit(code=1)

    mvm_cli.success(f"Created: {config.name}")
    mvm_cli.info(f"  SUBNET:    {config.subnet}")
    mvm_cli.info(f"  IPv4 Gateway: {config.ipv4_gateway}")
    mvm_cli.info(f"  Bridge:  {config.bridge}")
    mvm_cli.info(f"  NAT:     {'True' if config.nat_enabled else 'False'}")
    if config.nat_gateways:
        mvm_cli.info(f"  NAT gateways: {', '.join(config.nat_gateways_list)}")


@network_app.command(
    name="rm",
    context_settings={"allow_extra_args": True, "ignore_unknown_options": True},
)
@handle_errors
def network_rm(
    names: list[str] = typer.Argument(
        None,
        help="Network names to remove",
        autocompletion=_complete_network_names,
    ),
    force: bool = typer.Option(
        False, "--force", "-f", help="Remove even if referenced by VMs"
    ),
) -> None:
    """Remove one or more networks by name."""
    effective_names = list(names) if names else []
    if not effective_names:
        mvm_cli.error("Provide at least one network name")
        raise typer.Exit(code=1)

    result = NetworkOperation.remove(
        NetworkInput(name=effective_names), force=force
    )
    if result.status in ("error", "failure"):
        mvm_cli.error(result.message)
        raise typer.Exit(code=1)
    for name in effective_names:
        mvm_cli.success(f"Removed: {name}")


@network_app.command(
    name="inspect",
    context_settings={"allow_extra_args": True, "ignore_unknown_options": True},
)
@handle_errors
def network_inspect(
    ctx: typer.Context,
    name: str | None = typer.Argument(
        None,
        help="Network name",
        autocompletion=_complete_network_names,
    ),
    json_output: bool = typer.Option(False, "--json", help="Output as JSON"),
    tree: bool = typer.Option(False, "--tree", help="Output in tree format"),
) -> None:
    """Show detailed information about a network."""
    name = mvm_cli.check_name_arg(ctx, name)
    info = NetworkOperation.inspect(NetworkInput(name=[name]))

    if json_output:
        typer.echo(json.dumps(info, indent=2, default=str))
        return

    net_name = info.get("network", {}).get("name", name)
    mvm_cli.print_dict_tree(info, title=f"Network: {net_name}")


@network_app.command(
    name="sync",
    context_settings={"allow_extra_args": True, "ignore_unknown_options": True},
)
@handle_errors
def network_sync(
    ctx: typer.Context,
    identifier: str | None = typer.Argument(
        None,
        help="Network name or ID (omit for all)",
        autocompletion=_complete_network_names,
    ),
    json_output: bool = typer.Option(False, "--json", help="Output as JSON"),
) -> None:
    """Sync iptables rules between database and host."""
    network_id: str | None = None
    if identifier is not None:
        network = NetworkOperation.get(NetworkInput(name=[identifier]))
        network_id = network.id

    sync_result = NetworkOperation.sync(network_id)
    if sync_result.status in ("error", "failure"):
        mvm_cli.error(sync_result.message)
        raise typer.Exit(code=1)
    results = sync_result.item
    if results is None:
        mvm_cli.error("Sync returned no results")
        raise typer.Exit(code=1)

    if json_output:
        typer.echo(json.dumps(results, indent=2))
        return

    # Build a name map for all networks to avoid N+1 lookups
    all_networks = NetworkOperation.list_all()
    name_map = {n.id: n.name for n in all_networks}

    rows = []
    for nid, counts in results.items():
        short_id = mvm_cli.format_id(nid)
        name = name_map.get(nid, nid[:8])
        rows.append(
            [
                short_id,
                name,
                str(counts["verified"]),
                str(counts["added"]),
                str(counts["orphaned"]),
            ]
        )

    mvm_cli.table(
        columns=["ID", "Name", "Verified", "Added", "Orphaned"],
        rows=rows,
    )

File: internal/cli/network.go
package cli

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"mvmctl/internal/cli/common"
	"mvmctl/internal/lib/crypto"
	"mvmctl/internal/lib/model"
	libnet "mvmctl/internal/lib/network"
	"mvmctl/pkg/api"
	"mvmctl/pkg/api/inputs"

	"github.com/spf13/cobra"
)

// networkColumns defines the local listing columns for networks.
var networkColumns = []common.ListingColumn{
	{Header: "", Extract: func(v any) string { return common.Cli.FormatMarker(v.(*model.NetworkItem).IsDefault) }},
	{Header: "ID", Extract: func(v any) string { return common.Cli.FormatID(v.(*model.NetworkItem).ID) }},
	{Header: "Name", Extract: func(v any) string {
		return common.Cli.FormatName(v.(*model.NetworkItem).Name, !v.(*model.NetworkItem).IsPresent)
	}},
	{Header: "Subnet", Extract: func(v any) string { return v.(*model.NetworkItem).Subnet }},
	{Header: "NAT", Extract: func(v any) string {
		if v.(*model.NetworkItem).NATEnabled {
			return "yes"
		}
		return "no"
	}},
	{Header: "Bridge", Extract: func(v any) string { return v.(*model.NetworkItem).Bridge }, LongOnly: true},
	{Header: "VMs", Extract: func(v any) string {
		l := v.(*model.NetworkItem).Leases
		if l != nil {
			return fmt.Sprintf("%d", len(l))
		}
		return "0"
	}, LongOnly: true},
	{
		Header:  "Created",
		Extract: func(v any) string { return common.Cli.FormatTimestamp(v.(*model.NetworkItem).CreatedAt, "relative") },
	},
}

// NewNetworkCmd creates the network command and its subcommands.
func NewNetworkCmd(networkAPI api.NetworkAPI, configAPI api.ConfigAPI) *cobra.Command {
	cmd := &cobra.Command{
		Use:     "network",
		Aliases: []string{"net"},
		Short:   "Network management",
	}

	cmd.AddCommand(newNetworkListCmd(networkAPI, configAPI))
	cmd.AddCommand(newNetworkCreateCmd(networkAPI))
	cmd.AddCommand(newNetworkRemoveCmd(networkAPI))
	cmd.AddCommand(newNetworkInspectCmd(networkAPI))
	cmd.AddCommand(newNetworkSyncCmd(networkAPI))
	cmd.AddCommand(newNetworkDefaultCmd(networkAPI))

	return cmd
}

func newNetworkListCmd(networkAPI api.NetworkAPI, configAPI api.ConfigAPI) *cobra.Command {
	var jsonOutput bool
	var longOutput bool

	cmd := &cobra.Command{
		Use:     "ls",
		Aliases: []string{"list"},
		Short:   "List all networks.",
		RunE: func(cmd *cobra.Command, args []string) error {
			nets, err := networkAPI.NetworkListAll(cmd.Context())
			if err != nil {
				return err
			}

			if jsonOutput {
				if nets == nil {
					nets = []*model.NetworkItem{}
				}
				data, _ := json.MarshalIndent(nets, "", "  ")
				fmt.Println(string(data))
				return nil
			}

			// Resolve listing style from --long flag or DB config
			style := common.Cli.ResolveListingStyle(cmd.Context(), configAPI, longOutput)
			common.RenderListing(nets, networkColumns, style)
			return nil
		},
	}

	cmd.Flags().BoolVar(&jsonOutput, "json", false, "Output as JSON")
	cmd.Flags().BoolVar(&longOutput, "long", false, "Show full listing with all columns")
	return cmd
}

func newNetworkCreateCmd(networkAPI api.NetworkAPI) *cobra.Command {
	var subnet string
	var ipv4Gateway string
	var noNAT bool
	var natGateways string
	var nonInteractive bool
	var setDefault bool

	cmd := &cobra.Command{
		Use:   "create [name]",
		Short: "Create a named network.",
		Args:  cobra.MaximumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			var name string
			if len(args) > 0 {
				var err error
				name, err = common.Cli.CheckArg(cmd, args[0])
				if err != nil {
					return err
				}
			} else {
				// Let Cobra handle the missing argument error
				return fmt.Errorf("missing required argument: name")
			}

			if subnet == "" {
				return fmt.Errorf("Missing required option '--subnet'")
			}

			if natGateways == "" && !noNAT && !nonInteractive {
				var err error
				natGateways, err = resolveUserNATGateways(cmd.Context())
				if err != nil {
					return err
				}
			}

			var natGatewaysList []string
			if natGateways != "" {
				for _, g := range strings.Split(natGateways, ",") {
					g = strings.TrimSpace(g)
					if g != "" {
						natGatewaysList = append(natGatewaysList, g)
					}
				}
			}

			var gw *string
			if ipv4Gateway != "" {
				gw = &ipv4Gateway
			}
			input := inputs.NetworkCreateInput{
				Name:        name,
				Subnet:      subnet,
				IPv4Gateway: gw,
				NATEnabled:  !noNAT,
				NATGateways: natGatewaysList,
				SetDefault:  setDefault,
			}

			net, err := networkAPI.NetworkCreate(cmd.Context(), input)
			if err != nil {
				return fmt.Errorf("create network failed: %w", err)
			}

			common.Cli.PrintDictTree(common.Cli.ToMap(net), fmt.Sprintf("Network: %s", net.Name))

			if setDefault {
				common.Cli.Success(fmt.Sprintf("Default network set to: %s", net.Name))
			}

			return nil
		},
	}

	cmd.Flags().StringVar(&subnet, "subnet", "", "IP subnet in SUBNET notation (e.g. 192.168.100.0/24)")
	cmd.Flags().StringVar(&ipv4Gateway, "ipv4-gateway", "", "Gateway IPv4 for the bridge")
	cmd.Flags().BoolVar(&noNAT, "no-nat", false, "Disable NAT/masquerade")
	cmd.Flags().
		StringVar(&natGateways, "nat-gateways", "", "Physical interfaces for NAT (comma-separated, auto-detected if not provided)")
	cmd.Flags().
		BoolVar(&nonInteractive, "non-interactive", false, "Skip interactive prompts (auto-detect NAT interfaces)")
	cmd.Flags().BoolVarP(&setDefault, "default", "d", false, "Set as default network")
	return cmd
}

// resolveUserNATGateways prompts the user to select NAT gateway interfaces.
func resolveUserNATGateways(ctx context.Context) (string, error) {
	interfaces, err := libnet.GetPhysicalInterfaces()
	if err != nil {
		return "", fmt.Errorf("failed to list network interfaces: %w", err)
	}
	if len(interfaces) == 0 {
		return "", fmt.Errorf("no network interfaces found")
	}
	if len(interfaces) == 1 {
		return interfaces[0], nil
	}
	selected, err := common.Cli.PromptMultiSelect(
		ctx,
		"Select interface(s) for NAT (internet access):",
		interfaces,
		nil,
	)
	if err != nil {
		return "", err
	}
	return strings.Join(selected, ","), nil
}

func newNetworkRemoveCmd(networkAPI api.NetworkAPI) *cobra.Command {
	var force bool

	cmd := &cobra.Command{
		Use:               "rm [selectors...]",
		Aliases:           []string{"remove", "delete", "del"},
		Short:             "Remove one or more networks by name.",
		Args:              cobra.ArbitraryArgs,
		ValidArgsFunction: completeNetworkNames,
		RunE: func(cmd *cobra.Command, args []string) error {
			if len(args) == 0 {
				return fmt.Errorf("usage error")
			}

			err := networkAPI.NetworkRemove(cmd.Context(), inputs.NetworkInput{Identifiers: args}, force)
			if err != nil {
				return fmt.Errorf("remove failed: %w", err)
			}

			for _, name := range args {
				common.Cli.Success(fmt.Sprintf("Removed: %s", name))
			}

			return nil
		},
	}

	cmd.Flags().BoolVarP(&force, "force", "f", false, "Remove even if referenced by VMs")
	return cmd
}

func newNetworkInspectCmd(networkAPI api.NetworkAPI) *cobra.Command {
	var jsonOutput bool

	cmd := &cobra.Command{
		Use:               "inspect [selector]",
		Short:             "Show detailed information about a network.",
		Args:              cobra.MaximumNArgs(1),
		ValidArgsFunction: completeNetworkNames,
		RunE: func(cmd *cobra.Command, args []string) error {
			var name string
			if len(args) > 0 {
				var err error
				name, err = common.Cli.CheckArg(cmd, args[0])
				if err != nil {
					return err
				}
			} else {
				fmt.Fprintf(os.Stderr, "Error: missing required argument\n")
				return fmt.Errorf("missing required argument")
			}

			info, err := networkAPI.NetworkInspect(cmd.Context(), inputs.NetworkInput{Identifiers: []string{name}})
			if err != nil {
				return fmt.Errorf("network not found: %s", name)
			}

			if jsonOutput {
				fmt.Println(common.MarshalJSONDefaultStr(info))
				return nil
			}

			common.Cli.PrintDictTree(common.Cli.ToMap(info), fmt.Sprintf("Network: %s", info.Network.Name))
			return nil
		},
	}

	cmd.Flags().BoolVar(&jsonOutput, "json", false, "Output as JSON")
	return cmd
}

func newNetworkSyncCmd(networkAPI api.NetworkAPI) *cobra.Command {
	var jsonOutput bool

	cmd := &cobra.Command{
		Use:               "sync [selectors...]",
		Short:             "Sync iptables rules between database and host.",
		Args:              cobra.ArbitraryArgs,
		ValidArgsFunction: completeNetworkNames,
		RunE: func(cmd *cobra.Command, args []string) error {
			var netInput inputs.NetworkInput
			if len(args) > 0 {
				netInput = inputs.NetworkInput{Identifiers: args}
			}
			results, err := networkAPI.NetworkSync(cmd.Context(), netInput)
			if err != nil {
				return fmt.Errorf("sync failed: %w", err)
			}

			if jsonOutput {
				fmt.Println(common.MarshalJSONDefaultStr(results))
				return nil
			}

			// Build name map from all networks
			allNetworks, listErr := networkAPI.NetworkListAll(cmd.Context())
			if listErr != nil {
				return listErr
			}
			nameMap := make(map[string]string)
			for _, n := range allNetworks {
				nameMap[n.ID] = n.Name
			}

			// Build table rows
			rows := make([][]string, 0, len(results))
			for nid, counts := range results {
				shortID := common.Cli.FormatID(nid)
				name := nameMap[nid]
				if name == "" {
					name = crypto.Truncate(nid, 8)
				}
				rows = append(rows, []string{
					shortID,
					name,
					fmt.Sprintf("%d", counts["verified"]),
					fmt.Sprintf("%d", counts["added"]),
					fmt.Sprintf("%d", counts["orphaned"]),
				})
			}

			common.Cli.Table([]string{"ID", "Name", "Verified", "Added", "Orphaned"}, rows)
			return nil
		},
	}

	cmd.Flags().BoolVar(&jsonOutput, "json", false, "Output as JSON")
	return cmd
}

func newNetworkDefaultCmd(networkAPI api.NetworkAPI) *cobra.Command {
	cmd := &cobra.Command{
		Use:   "default [selector]",
		Short: "Set a network as the default for VM creation.",
		Args:  cobra.MaximumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			var name string
			if len(args) > 0 {
				var err error
				name, err = common.Cli.CheckArg(cmd, args[0])
				if err != nil {
					return err
				}
			} else {
				fmt.Fprintf(os.Stderr, "Error: missing required argument\n")
				return fmt.Errorf("missing required argument")
			}

			err := networkAPI.NetworkSetDefault(cmd.Context(), inputs.NetworkInput{Identifiers: []string{name}})
			if err != nil {
				return fmt.Errorf("set default failed: %w", err)
			}

			common.Cli.Success(fmt.Sprintf("Default network set to: %s", name))
			return nil
		},
	}
	return cmd
}
